test/easystartest.js

```diff
@@ -1,349 +1,411 @@
+        easyStar.setAcceptableTiles([0]);
+        easyStar.setDirectionalCondition(2, 1, [EasyStar.TOP]);
+        easyStar.setDirectionalCondition(1, 1, [EasyStar.RIGHT]);
+        easyStar.setDirectionalCondition(0, 1, [EasyStar.RIGHT]);
+        easyStar.setDirectionalCondition(0, 0, [EasyStar.BOTTOM]);
+
+        easyStar.findPath(2, 0, 0, 0, function (path) {
+            expect(path).not.toBeNull();
+            expect(path.length).toEqual(5);
+            expect(path[2]).toEqual({x: 1, y: 1});
+            done();
+        });
+
+        easyStar.calculate();
     });
 
-    easyStar.calculate();
-  })
-
-  it("It should handle tiles with a directional condition", function (done) {
-      var easyStar = new EasyStar.js();
-      var grid = [
-          [0, 1, 0],
-          [0, 0, 0],
-          [0, 0, 0],
-      ];
-      easyStar.setGrid(grid);
-      easyStar.enableDiagonals();
-      easyStar.setAcceptableTiles([0]);
-      easyStar.setDirectionalCondition(2, 1, [EasyStar.TOP]);
-      easyStar.setDirectionalCondition(1, 2, [EasyStar.TOP_RIGHT]);
-      easyStar.setDirectionalCondition(2, 2, [EasyStar.LEFT]);
-      easyStar.setDirectionalCondition(1, 1, [EasyStar.BOTTOM_RIGHT]);
-      easyStar.setDirectionalCondition(0, 1, [EasyStar.RIGHT]);
-      easyStar.setDirectionalCondition(0, 0, [EasyStar.BOTTOM]);
-
-      easyStar.findPath(2, 0, 0, 0, function (path) {
-          expect(path).not.toBeNull();
-          expect(path.length).toEqual(7);
-          expect(path[3]).toEqual({ x: 2, y: 2})
-          done();
-      });
-
-      easyStar.calculate();
-  })
-
-  it("It should handle tiles with a directional condition and no corner cutting", function (done) {
-      var easyStar = new EasyStar.js();
-      easyStar.disableCornerCutting();
-      var grid = [
-          [0, 1, 0],
-          [0, 0, 0],
-          [0, 0, 0],
-      ];
-      easyStar.setGrid(grid);
-      easyStar.enableDiagonals();
-      easyStar.setAcceptableTiles([0]);
-      easyStar.setDirectionalCondition(2, 1, [EasyStar.TOP]);
-      easyStar.setDirectionalCondition(1, 1, [EasyStar.RIGHT]);
-      easyStar.setDirectionalCondition(0, 1, [EasyStar.RIGHT]);
-      easyStar.setDirectionalCondition(0, 0, [EasyStar.BOTTOM]);
-
-      easyStar.findPath(2, 0, 0, 0, function (path) {
-          expect(path).not.toBeNull();
-          expect(path.length).toEqual(5);
-          expect(path[2]).toEqual({ x: 1, y: 1})
-          done();
-      });
-
-      easyStar.calculate();
-  })
+    it('It should invoke the findPath callback immediately when sync mode is enabled', function () {
+        var easyStar = new EasyStar.js();
+        var result = [];
+        var grid = [
+            [0, 1, 0],
+            [0, 0, 0],
+            [0, 0, 0],
+        ];
+
+        easyStar.enableSync();
+        easyStar.setGrid(grid);
+        easyStar.setAcceptableTiles([0]);
+
+        easyStar.findPath(0, 0, 1, 1, function (path) {
+            result.push.apply(result, path);
```

Comment:
When you revert the Zalgo bug this test won't work anymore. I would recommend instead to set the `iterationsPerCalculation` value to be value low (like 1), a single `calculate()` call should still find the path.

Reply:
Interestingly this test does still work with the zalgo bug reverted. So I understand, when would you expect array referenced  by `result` to be updated in this context?

This is the `enableSync` test, so our callback is immediately invoked on the (I assume) current execution stack. Therefore our result will have changed before `calculate` has returned, as calculate is synchronous, and our assertion takes place after `calculate`.

Reply:
My expectation would be that if you reverted the change to the `setTimeout`, that would `result` would be an empty array until we enter the callback, because the callback is executed only after the assertions below calculate.